src/ln/channelmanager.rs

```diff
@@ -1463,56 +1487,79 @@ impl ChannelManager {
 		} else { false }
 	}
 	fn claim_funds_internal(&self, mut channel_state_lock: MutexGuard<ChannelHolder>, source: HTLCSource, payment_preimage: PaymentPreimage) {
-		match source {
-			HTLCSource::OutboundRoute { .. } => {
-				mem::drop(channel_state_lock);
-				let mut pending_events = self.pending_events.lock().unwrap();
-				pending_events.push(events::Event::PaymentSent {
-					payment_preimage
-				});
-			},
-			HTLCSource::PreviousHopData(HTLCPreviousHopData { short_channel_id, htlc_id, .. }) => {
-				//TODO: Delay the claimed_funds relaying just like we do outbound relay!
-				let channel_state = channel_state_lock.borrow_parts();
-
-				let chan_id = match channel_state.short_to_id.get(&short_channel_id) {
-					Some(chan_id) => chan_id.clone(),
-					None => {
-						// TODO: There is probably a channel manager somewhere that needs to
-						// learn the preimage as the channel already hit the chain and that's
-						// why its missing.
-						return
-					}
-				};
+		let (their_node_id, err) = loop {
```

Comment:
Do a loop is really needed ? Why instead not use a temp Option value to catch a potential error ?

Reply:
handle_monitor_err will move out of the chan field, so doing anything other than a loop ends you up in borrow checker hell.

Reply:
Ok